File: StriverA2Z/C++/3_ArraysQuestion/3.3_Hard/largest_subarray_with_zero_sum.cpp
#include<bits/stdc++.h>

using namespace std;

void largestSubarrayZeroSum(int arr[], int n) {
    map<int, int> mpp;
    int maxi = 0;
    int sum = 0;

    for (int i = 0; i < n; i++) {
        sum += arr[i];
        if (sum == 0) {
            maxi = i + 1;
        } else {
            if (mpp.find(sum) != mpp.end()) {
                maxi = max(maxi, i - mpp[sum]);
            } else {
                mpp[sum] = i;
            }
        }
    }

    cout << maxi;
}

int main() {
    int arr[] = {1, -1, 0, 0, 2};
    largestSubarrayZeroSum(arr, sizeof(arr) / sizeof(arr[0]));
    return 0;
}

File: StriverA2Z/Java/Arrays/Hard/LargestSubarrayWithZeroSum.java
import java.util.HashMap;

public class LargestSubarrayWithZeroSum {
    public static void main(String[] args) {
        int[] arr = {1, -1, 0, 0, 2};
        largestSubarrayZeroSum(arr);
    }

    static void largestSubarrayZeroSum(int[] arr) {
        HashMap<Integer, Integer> mpp = new HashMap<>();
        int maxi = 0;
        int sum = 0;
        for (int i = 0; i < arr.length; i++) {
            sum += arr[i];
            if (sum == 0) {
                maxi = i + 1;
            } else {
                if (mpp.get(sum) != null) {
                    maxi = Math.max(maxi, mpp.get(sum));
                } else {
                    mpp.put(sum, i);
                }
            }
        }
        System.out.println(maxi);
    }
}
